pkg/deployment/config.go

```diff
@@ -116,10 +121,53 @@ func defaultConfig() *Config {
 	return cfg
 }
 
+func (d Config) PodRuntimeClassName(lbs map[string]string) *string {
+	runtimeClassName := ""
+	specificity := -1
+	for k, v := range d.RuntimeClassNames {
+		if !v.Matches(lbs) || v.specificity() < specificity {
+			continue
+		}
+		if v.specificity() > specificity || strings.Compare(k, runtimeClassName) < 0 {
+			runtimeClassName = k
+			specificity = v.specificity()
+		}
+	}
+	if runtimeClassName == "" {
+		return nil
+	}
+	return ptr.String(runtimeClassName)
```

Comment:
Do we want to validate with https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L314 and fail fast?

Reply:
@skonto, thank you for your thoughts. More validation is essential and will make it more robust.

I've tried adding in
```
+			if err := apivalidation.ValidateRuntimeClassName(class, field.NewPath(".data."+RuntimeClassNameKey)); err != nil {
+				return nil, fmt.Errorf("%v invalid: %v", RuntimeClassNameKey, err)
+			}
```

but now have an import failure. Haven't gotten one of these errors in a bit
```
$ go mod tidy
go: finding module for package k8s.io/kubernetes/pkg/apis/core/validation
go: found k8s.io/kubernetes/pkg/apis/core/validation in k8s.io/kubernetes v1.30.2
go: downloading k8s.io/component-helpers v0.0.0
go: downloading k8s.io/kubelet v0.0.0
go: downloading k8s.io/cloud-provider v0.0.0
go: finding module for package k8s.io/client-go/features
go: found k8s.io/client-go/features in k8s.io/client-go v0.30.2
go: knative.dev/serving/pkg/deployment imports
        k8s.io/kubernetes/pkg/apis/core/validation imports
        k8s.io/component-helpers/node/util/sysctl: reading k8s.io/component-helpers/go.mod at revision v0.0.0: unknown revision v0.0.0
go: knative.dev/serving/pkg/deployment imports
        k8s.io/kubernetes/pkg/apis/core/validation imports
        k8s.io/component-helpers/scheduling/corev1: reading k8s.io/component-helpers/go.mod at revision v0.0.0: unknown revision v0.0.0
go: knative.dev/serving/pkg/deployment imports
        k8s.io/kubernetes/pkg/apis/core/validation imports
        k8s.io/kubelet/pkg/apis: reading k8s.io/kubelet/go.mod at revision v0.0.0: unknown revision v0.0.0
go: knative.dev/serving/pkg/deployment imports
        k8s.io/kubernetes/pkg/apis/core/validation imports
        k8s.io/kubernetes/pkg/cluster/ports imports
        k8s.io/cloud-provider/options: reading k8s.io/cloud-provider/go.mod at revision v0.0.0: unknown revision v0.0.0
```

from adding the dependencies
```
	"k8s.io/apimachinery/pkg/util/validation/field"
	apivalidation "k8s.io/kubernetes/pkg/apis/core/validation"
```

Reply:
You can just import `apimachineryvalidation.NameIsDNSSubdomain(name, false)` and it should be fine.

Reply:
updated in 4adbced